keps/sig-storage/20190408-volume-scheduling-limits.md

```diff
@@ -0,0 +1,297 @@
+type VolumeLimits struct {
+	// Count is maximum number of volumes provided by the CSI driver that can be used by the node
+	// "nil" represents no limits - the node can handle any number of volumes of the driver.
+	Count *int32 `json:"count,omitempty" protobuf:"varint,1,opt,name=count`
+
+	// Future proof: max. total size of volumes on the node can be added later
+}
+```
+
+CSINode example:
+
+```yaml
+apiVersion: storage.k8s.io/v1beta1
+kind: CSINode
+metadata:
+  name: ip-172-18-4-112.ec2.internal
+spec:
+  drivers:
+    - name: ebs.csi.aws.com
+      # Already existing fields
+      nodeID: ip-172-18-4-112.ec2.internal
+      topologyKeys:
+         # ...
+      # New API:
+      allocatable:
+        # AWS node can attach max. 40 volumes, 1 is reserved for the system
+        count: 39
+    - name: org.kernel.nfs
+      allocatable:
+        # NFS does not impose any limits of volumes mounted to the node
+        count:    # nil means "no limit"
+```
+
+### Implementation
+
+Section below describes behaviour of old predicates, CSI predicate and scheduler after the proposal has been implemented.
+For brevity - "old predicates" refers to now deprecated cloudprovider specific predicates - `MaxEBSVolumeCount`, `MaxGCEPDVolumeCount`, `MaxAzureDiskVolumeCount` and `MaxCinderVolumeCount`.
+
+#### Implementation Detail for all CSI Drivers
+
+* Kubelet will create `CSINode` instance during initial CSI Driver registration.
+  * Limits of each CSI volume plugin will be added to `CSINode.spec.drivers[xyz].allocatable`.
+  * User may NOT change `CSINode.spec.drivers[xyz].allocatable` to override volume plugin / CSI driver values, e.g. to "reserve" some attachment to the operating system. Kubelet will periodically reconcile `CSINode` and overwrite the value.
+    * Especially, `kubelet --kube-reserved` or `--system-reserved` cannot be used to "reserve" volumes for kubelet or the OS. It is not possible with existing kubelet and this KEP does not change it. We expect that CSI drivers will have configuration options / cmdline arguments to reserve some volumes and they will report their limit already reduced by that reserved amount.
+* Scheduler will respect `Node.status.allocatable` and `Node.status.capacity` for CSI volumes if `CSINode` object is not available or has missing entry in `CSINode.spec.drivers[xyz].allocatable` during a deprecation period but kubelet will stop populating `Node.status.allocatable` and `Node.status.capacity` for CSI volumes.
+  * After deprecation period for CSI volumes, limits coming from `Node.status.allocatable` and `Node.status.capacity` will be completely ignored by the scheduler.
+  * After deprecation period, scheduler won't schedule any pods that use CSI volumes to a node with missing `CSINode` instance or missing driver in `CSINode.Spec.Drivers`. It is expected that it happens only during node registration when `Node` exists and `CSINode` doesn't and it self-heals quickly.
```

Comment:
IMO, whole sentence "It is expected that it happens only during node registration when `Node` exists and `CSINode` doesn't and it self-heals quickly" is obsolete. There is no self-heal because there is no limits in CSINode created at kubelet startup. Scheduler needs to wait for a driver to get installed there.

Reply:
removed.